### ISSUE TITLE
HoeffdingTree predict_proba returns different dimensions per sample

### ISSUE THREAD

**Issue by lfleck** [scikit-multiflow/scikit-multiflow#46]
### Expected behaviour
output_dim = num_samples x num_classes

### Actual behaviour
output is an array of lists with different length

### Steps to reproduce the behaviour
test files are attached
[data.zip](https://github.com/scikit-multiflow/scikit-multiflow/files/2302548/data.zip)
```
import numpy as np
from skmultiflow.trees.hoeffding_tree import HoeffdingTree

data = np.load("test_data.npy")
labels = np.load("test_labels.npy")

model = HoeffdingTree()
model.partial_fit(data, labels)
pred_proba = model.predict_proba(data)
```
`>> [list([1.0]), ... , list([0, 1.0]), ... , list([1.0]) , ...] `

**Comment by jacobmontiel** [comment]
predict_proba() return the probabilities mapped into an array where the index corresponds to the class.

This is implemented in
https://github.com/scikit-multiflow/scikit-multiflow/blob/3bac7c8274948b9c7f4518994a85f4bb5a5b5e0e/src/skmultiflow/trees/hoeffding_tree.py#L1083

The size of the array depends on the max class value. This is useful when not all classes are known at the beginning of the stream.

What about using the `classes` attribute (if it is defined)?
```python
if self.classes is not None:
    y_proba = np.zeros(max(self.classes) + 1)
else:
    y_proba = [0] * (int(max(votes.keys())) + 1)
```

**Comment by lfleck** [comment]
Yes that would fix it. Currently, `self.classes` is not defined, 
https://github.com/scikit-multiflow/scikit-multiflow/blob/3bac7c8274948b9c7f4518994a85f4bb5a5b5e0e/src/skmultiflow/trees/hoeffding_tree.py#L936  
https://github.com/scikit-multiflow/scikit-multiflow/blob/3bac7c8274948b9c7f4518994a85f4bb5a5b5e0e/src/skmultiflow/trees/hoeffding_tree.py#L960
but it could be set in the first call of `partial_fit`

However, I would prefer 
```python
if self.classes is not None:
    y_proba = np.zeros(len(self.classes))
else:
    y_proba = np.zeros(len(votes.keys()))
```
since a 3-class classification problem could have for example labels in [2, 6, 8]. 
What do you think @jacobmontiel

**Comment by jacobmontiel** [comment]
Yes, we should set `self.classes` on the first call.

Now, we need `max(self.classes) + 1` to map correctly into the 0-based array given that the index implies the class. This is used in the `predict()` method:
https://github.com/scikit-multiflow/scikit-multiflow/blob/3bac7c8274948b9c7f4518994a85f4bb5a5b5e0e/src/skmultiflow/trees/hoeffding_tree.py#L1055

**Comment by lfleck** [comment]
Ah, I see. 
Than something like
```python
if self.classes is not None:
     prediction = self.classes[index]
```
would be possible. 
Alternative, a hint in the documentation mentioning that classes have to be zero-based would also be fine, as the user can apply e.g. `sklearn.preprocessing.LabelEncoder()` if needed.

**Comment by jacobmontiel** [comment]
The nice thing about the `max` trick is that the user does not need to transform the labels. For example, if classes are [2, 6, 8] we map to an array of size 9. But in most cases, this is not necessary given that classes usually are defined zero-based. Nonetheless, it is true that we need to document that zero-based classes are preferred. The worst case scenario will be a very sparse label set.

I have pushed a fix: 4aec5ffb973aebd6765a8af0e49101f4d22718bf

From your example:

```python
classes = np.unique(labels)
model = HoeffdingTree()
model.partial_fit(data, labels, classes=classes)
pred_proba = model.predict_proba(data)

>>> array([[1., 0.],
       [1., 0.],
       [1., 0.],
       ...,
       [0., 1.],
       [1., 0.],
       [1., 0.]])

```

**Comment by lfleck** [comment]
Perfect. Thank you very much
